### PR TITLE
Hash functions

### PR DESCRIPTION
**Pull request by SchrodingerZhu**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@ ifndef CC
 	CC = gcc
 endif
 ifndef CFLAGS
-	CFLAGS = -Wall -Wextra -O2 -g $(EXTRA_CFLAGS)
+	CFLAGS = -Wall -Wextra -pedantic -Werror -O2 -g $(EXTRA_CFLAGS)
```
Comment (chuigda) [inline Makefile:5]: Don't use `-Werror` anywhere under any circumstance.

File: src/cc_vec.c [hunk 2]
```diff
@@ -279,12 +279,12 @@ cc_error cc_vec_remove2(cc_vec *vec, cc_size idx, cc_size cnt) {
...
 static inline _Bool cc_vec_internal_remove_if_pred2(void *ptr, void *ctx) {
-    cc_pred *fn = (cc_pred*)ctx;
-    return fn(ptr);
+    cc_pred ** fn = (cc_pred**) ctx;
+    return (*fn)(ptr);
 }
 
 cc_error cc_vec_remove_if(cc_vec *vec, cc_pred pred) {
-    return cc_vec_remove_if2(vec, cc_vec_internal_remove_if_pred2, pred);
+    return cc_vec_remove_if2(vec, cc_vec_internal_remove_if_pred2, (void *)(&pred));
```
Comment (chuigda) [inline src/cc_vec.c:287]: why we are taking the address of a pointer?

File: src/include/cc_impl.h [hunk 3]
```diff
@@ -34,4 +34,8 @@
         } \
     })
 
+#define PROJECT_CL_UNUSED(X) ((void)(X))
+#define PROJECT_CL_IS_BIG_ENDIAN() (!*(unsigned char *)&(uint16_t){1})
```
Comment (chuigda) [inline src/include/cc_impl.h:38]: We can make it part of `cc_cfg.h`, and make assertion in `cc_defs.c`.

File: include/cc_hash.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#include <cc_defs.h>
+typedef struct {
```
Comment (chuigda) [inline include/cc_hash.h:2]: Why don't we make it more opaque?

File: include/cc_abort.h [hunk 5]
```diff
@@ -7,9 +7,9 @@
 extern "C" {
 #endif
 
-CC_ATTRIBUTE_COLD void cc_abort();
+CC_ATTRIBUTE_COLD void cc_abort(void);
 
-CC_ATTRIBUTE_COLD _Bool cc_dump_stacktrace();
+CC_ATTRIBUTE_COLD _Bool cc_dump_stacktrace(void);
```
Comment (chuigda) [inline include/cc_abort.h:12]: Would adding a `void` or not affect semantics of function declarations? I remembered that function declarations are actually more explicit than definitions...

**Review by chuigda — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (chuigda) [reply Makefile:5]: 让我来解释一下这一点：随着时间发展，编译器肯定会不断改进，新的警告类别会被源源不断添加进去。考虑到我们作为*高素质*的 Project-CL 开发人员，肯定不会对这些警告熟视无睹，所以把它标记为错误与否并没有关系。与此相反，使用 Project-CL 的人可能会在某天升级编译器之后突然发现项目没法编译了。虽然 `CFLAGS` 可以从外部予以变更，但一个项目对 Project-CL 的依赖可能是嵌套在很深的地方的，比如 A -> B -> C -> D -> Project-CL，也就是开发者要改项目 D 指定的 Project-CL 编译配置，这在很多构建系统里都非常困难。而在这种情况下开发者甚至可能会 Fork 一份 D，这就引起了维护困难和事实上的社区分裂。所以由我们 —— 一个库的维护者 —— 来指定 `-Werror` 是绝对不行的。

File: src/cc_vec.c (hunk 2 above)
Comment (SchrodingerZhu) [reply src/cc_vec.c:287]: casting function pointer to object pointer is incorrect in pedantic C

File: include/cc_abort.h (hunk 5 above)
Comment (SchrodingerZhu) [reply include/cc_abort.h:12]: function declaration without void is not allowed in pedantic C.

File: Makefile (hunk 1 above)
Comment (SchrodingerZhu) [reply Makefile:5]: agreed

**Comment by SchrodingerZhu** [comment]
Also, please let me add some little tests before merging

File: include/cc_hash.h (hunk 4 above)
Comment (SchrodingerZhu) [reply include/cc_hash.h:2]: do you mean a pointer? that would require allocs I suppose?

**Comment by SchrodingerZhu** [comment]
@chuigda tests passed at https://github.com/SchrodingerZhu/cclib/pull/1

**Comment by SchrodingerZhu** [comment]
hold on this. hash quality test failed. work on it.

**Comment by SchrodingerZhu** [comment]
@chuigda seems fixed but let me make the whole report

**Comment by SchrodingerZhu** [comment]
[cc-untable-hash-aarch64.txt](https://github.com/chuigda/Project-CL/files/11013397/cc-untable-hash-aarch64.txt)

**Comment by SchrodingerZhu** [comment]
[cc-stable-hash-aarch64.txt](https://github.com/chuigda/Project-CL/files/11013612/cc-stable-hash-aarch64.txt)

**Comment by SchrodingerZhu** [comment]
I think it is ready for merge now. I am not sure about the formatting you required. Do you have time to format and reorder the macros as you wish?

**Review by chuigda — APPROVED** [review]

**Pull request merged**